Implements Spade functionality, and adds the sublime project file.

Zerrien opened the pull request:
Spade: Per-input recorder to playback a user's coding session.

Implementation: Hooks into the SpellView's ACE and records passively. Compresses and saves this information to a user's session if the user has been on a level for more than 10 minutes, or they press the contact button. There is currently no way to play this back, but UTF16 strings should be properly recorded in a user's session under codeLog.

Currently includes Spade as a vendor script, but will be converted to a Bower module eventually.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -106,6 +107,21 @@ module.exports = class SpellView extends CocoView
     @initAutocomplete aceConfig.liveCompletion ? true
 
+    # Create a Spade to 'dig' into Ace.
+    if Spade
```
We are always building Spade now, right? We can get rid of the `if Spade` check. (It would need to be `if Spade?` anyway.)

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -106,6 +107,21 @@ module.exports = class SpellView extends CocoView
 
+    # Create a Spade to 'dig' into Ace.
+    if Spade
+      spade = @spade = new Spade()
```
Local variable not used; let's get rid of it.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -106,6 +107,21 @@ module.exports = class SpellView extends CocoView
+      # If a user is taking longer than 10 minutes, let's log it.
+      tenMin = 10 * 60 * 1000
+      view = @
+      @spadeWatch = setTimeout(->
```
Better style and garbage collection like this:

```
    spadeSaveDelay = 10 * 60 * 1000
    @spadeSaveTimeout = setTimeout @putSpade, spadeSaveDelay

  saveSpade: =>
    return if @destroyed
```

Call it `saveSpade` instead of `putSpade`? Also in the view's `destroy` method, clear the `@spadeSaveTimeout`.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -222,6 +238,7 @@ module.exports = class SpellView extends CocoView
         if not disableSpaces or (_.isNumber(disableSpaces) and disableSpaces < me.level())
           return @ace.execCommand 'insertstring', ' '
         line = @aceDoc.getLine @ace.getCursorPosition().row
+        console.log(@singleLineCommentRegex().test line)
```
Remember to check for debug logs like this before you commit.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +711,30 @@ module.exports = class SpellView extends CocoView
+  putSpade: () ->
+    # This happens when 1) The user has taken longer than 10 minutes, 2) the Contact button is pressed.
+    # Let's check to make sure SPADE exists.
+    if @spade
```
Just do `return unless @spade` so that we don't have an extra level of indentation and so that the "happy path" ("golden path") of the default case (we do have `@spade`) is the "normal", un-indented one, and exceptions are the indentations.

... but also I think we will now always have `@spade`, right? So we don't need this check either.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +711,30 @@ module.exports = class SpellView extends CocoView
+      # No need to write if there was absolutely nothing recorded.
+      if(condensedEvents.length is 0)
+        console.warn('SPADE saw no events, nothing to record')
+        return
```
We don't need a comment, a log, and a self-explanatory if-statement all saying the same thing. Also, don't check against 0 and don't put parens in CoffeeScript conditionals. We can also return if spadeEvents is empty instead of trying to condense first. What about just–

`return unless spadeEvents.length`

Also, the since we are calling these `condensedEvents`, compress/uncompress in Spade itself should be called condense/expand or something, and the following `compressedEvents` logic should probably move into Spade itself.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +711,30 @@ module.exports = class SpellView extends CocoView
+      @options.session.set('codeLog', compressedEvents)
+      @options.session.save()
+  
+  onShowVictory: () ->
```
Style: don't use parens for empty function arguments.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +711,30 @@ module.exports = class SpellView extends CocoView
+  
+  onShowVictory: () ->
+    if @spadeWatch
+      window.clearTimeout(@spadeWatch)
```
Also null out the timer variable after clearing it.

nwinter commented on vendor/scripts/spade.js:
```diff
@@ -0,0 +1,230 @@
+			keyHook = _elem;
+		}
+		keyHook.addEventListener("keydown", function(_event) {spade.createEvent(spade.target)});
+		//Maybe this is needed depending on Firefox/other browsers? Duplicate non-diff events get compiled down.
```
Have you tried just using `keypress` instead of `keydown` and `keyup`?

Zerrien replied:
Yes, not sure why it doesn't work, but it seemed to sometimes lose keypresses when typing.

nwinter commented on vendor/scripts/spade.js:
```diff
@@ -0,0 +1,230 @@
+		}
+		return compiledStack;
+	},
+	play: function(_stack, _elem) {
```
Do we as yet have a way to call this and play back the events?

Zerrien replied:
As of the latest commits we do, under /admin/codelogs

nwinter commented on vendor/scripts/spade.js:
```diff
@@ -0,0 +1,230 @@
+			});
+		}
+	},
+	clear: function() {
```
This is never called? Not sure why it is needed.

Also, we need a `destroy()` or `untrack()` method that we can call to remove our key and mouse listeners and possibly the playback timeout if we are playing, and then to call this when we stop tracking in SpellView or when SpellView is destroyed.

Zerrien commented:
Lots of fixes since last month:

Code logs are now stored in their own database table
Player sessions are given an array of code log ids
There is an admin debug page to watch the code log

nwinter commented on app/views/admin/CodeLogsView.coffee:
```diff
@@ -0,0 +1,46 @@
+    tooltip.style.borderRadius = "8px"
+    tooltip.id = "codelogs-tooltip"
+    tooltip.addEventListener 'blur', @onBlurTooltip
+    document.body.appendChild tooltip
```
We usually use jQuery to do DOM manipulation and event listeners like this. Not saying you have to, but for consistency, maybe better? Also I don't think the memory management here is enough; maybe if you add a `destroy` method (like most views have) that will get rid of any remaining tooltips. And do we need to remove the event listener ourselves or not when removing it from the DOM?

We should probably have this tooltip have a CSS class that we can style in our normal stylesheet, independent of the positioning bit (which could stay here).

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -47,6 +48,8 @@ module.exports = class SpellView extends CocoView
     'tome:maximize-toggled': 'onMaximizeToggled'
     'script:state-changed': 'onScriptStateChange'
     'playback:ended-changed': 'onPlaybackEndedChanged'
+    'level:contact-button-pressed': 'saveSpade'
```
For style consistency, we should probably call this `onContactButtonPressed` and then inside that method we can call the `saveSpade` method.

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -106,6 +108,15 @@ module.exports = class SpellView extends CocoView
+    @spade = new Spade()
+    @spade.track(@ace)
+
+    # If a user is taking longer than 10 minutes, let's log it.
```
Let's `return if @session.get('creator') isnt me.id or @session.fake` so that we don't try to save sessions when observing them or when a teacher is playing without saving a real session.

nwinter commented on server/codelogs/codelog_handler.coffee:
```diff
@@ -0,0 +1,24 @@
+    codelog.set 'userName', req.body.userName
+    codelog.set 'log', req.body.log
+    codelog.set 'date', new Date().toISOString()
+    codelog.set 'id', req.body.id
```
What's this bit about setting `id` here and on line 9? I think that Mongo will set the `_id`, right, and it's `_id`, not `id`.

Also I think we need to take out `date`, that's now automatically set as `created` by the Mongoose schema.

nwinter commented on server/codelogs/codelog_handler.coffee:
```diff
@@ -0,0 +1,24 @@
+    return codelog
+
+  post: (req, res) ->
+    super(req, res)
```
Don't really need this `post` I think, if it's not going to do anything.

sderickson commented on app/models/CodeLog.coffee:
```diff
@@ -0,0 +1,8 @@
+  urlRoot: '/db/codelogs'
+  @schema: require 'schemas/models/codelog.schema'
+  initialize: ->
+    super()
```
This initialize can be removed

sderickson commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +705,41 @@ module.exports = class SpellView extends CocoView
     return if @destroyed
     Backbone.Mediator.publish 'tome:hide-problem-alert', {}
 
+  saveSpade: (e) =>
```
Fat arrow unnecessary, Backbone binds event callbacks to the view automatically.

nwinter replied:
But this is also called via the `setTimeout` on line 117, which doesn't.

sderickson commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +705,41 @@ module.exports = class SpellView extends CocoView
+      userName: @options.session.get 'creatorName'
+      log: compressedEvents
+    })
+    codelog.on('save:success', (e) ->
```
`e` should be `codelog`, since that's what the first argument is. `e` is for "event", for jquery event callbacks and such.

sderickson commented on app/views/admin/CodeLogsView.coffee:
```diff
@@ -0,0 +1,46 @@
+    @codelogs = new CocoCollection([],
+      url: '/db/codelogs'
+      model: CodeLog
+    )
```
Make a CodeLogs collection in `/app/collections` and use that instead.

sderickson commented on server/codelogs/codelog_handler.coffee:
```diff
@@ -0,0 +1,24 @@
+  post: (req, res) ->
+    super(req, res)
+
+module.exports = new CodeLogHandler()
```
Handlers are deprecated. Instead, add routes to [`server/routes/index.coffee`](https://github.com/codecombat/codecombat/blob/master/server/routes/index.coffee), using middleware. You just need to GET a collection of them and POST new ones? Do something like [Article's routes here](https://github.com/codecombat/codecombat/blob/master/server/routes/index.coffee#L29-L30). You'll need to add `editableProperties` to the server CodeLog model, like [here](https://github.com/codecombat/codecombat/blob/master/server/models/Article.coffee#L39-L40). And there should be server tests like for [Article GET](https://github.com/codecombat/codecombat/blob/master/spec/server/functional/article.spec.coffee#L10-L27) and [Article POST](https://github.com/codecombat/codecombat/blob/master/spec/server/functional/article.spec.coffee#L121-L139). If you have any questions about this server stuff, let me know!

nwinter commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -694,6 +705,41 @@ module.exports = class SpellView extends CocoView
+    condensedEvents = @spade.condense(spadeEvents)
+    return unless condensedEvents.length
+    compressedEvents = LZString.compressToUTF16(JSON.stringify(condensedEvents))
+    codelog = new CodeLog({
```
Probably `codeLog` instead of `codelog` for case consistency.

Zerrien commented:
Another update, ready for review!

Switched from handler logic to async fanciness. Cleaned up a lot of the code based on the suggestions. Fixed file structure and locations.

sderickson commented on app/views/admin/CodeLogsView.coffee:
```diff
@@ -0,0 +1,49 @@
+    @spade = new Spade()
+    @codelogs = new CodeLogCollection()
+    @listenTo(@codelogs, 'sync', @onThangsLoaded)
+    @supermodel.loadCollection(@codelogs, 'codelogs')
```
Use `initialize` instead of `constructor`.

I think listening to `onThangsLoaded` should be unnecessary. If the supermodel is tracking requests, it'll render the whole view once everything has successfully loaded.

We're moving away from `loadCollection` and alike (so many things in flux). Instead, use `@supermodel.trackRequest(@codelogs.fetch())`. That way you're triggering the collection request directly, rather than indirectly through the supermodel.

sderickson commented on app/views/play/level/tome/SpellView.coffee:
```diff
@@ -705,6 +718,42 @@ module.exports = class SpellView extends CocoView
+        codelogs = [codelog.id]
+      @options.session.set('codeLogs', codelogs)
+      @options.session.save()
+    , @)
```
This would be better to do on the server side, as it would be more consistent, performant and succinct.

1. Create a middleware file for codelogs in `/server/middleware`
1. Copy the `rest` post middleware and make it specific to CodeLog.
1. Add logic to push the ID of the newly created `codeLog`'s id onto the session: `yield LevelSession.update({_id: mongoose.Types.ObjectId(req.body.sessionID)}, {$push: {codeLogs: codeLog._id}})`
1. Have `routes/index.coffee` use the custom middleware instead of the generic rest middleware for POST.

sderickson commented on server/models/CodeLog.coffee:
```diff
@@ -0,0 +1,24 @@
+  created:
+    type: Date
+    'default': Date.now
+}, {strict: false, read: config.mongo.readpref})
```
You should add userID and sessionID to the schema here. This will make Mongoose automagically convert strings ids to MongoDB ObjectIds.

sderickson commented on server/middleware/codelogs.coffee:
```diff
@@ -0,0 +1,22 @@
+      {$push:{codeLogs: doc._id}}
+    )
+
+    res.status(201).send(doc.toObject())
```
Should be made specific.
* `Model` -> `CodeLog`
* `doc` -> `codeLog`
* `post: (Model, options={}) -> wrap (req, res) ->` becomes `post: wrap (req, res) ->`
* In index.coffee: `codeLogs.post(CodeLog)` -> `codeLogs.post`

sderickson commented on server/models/LevelSession.coffee:
```diff
@@ -81,7 +81,7 @@ LevelSessionSchema.pre 'save', (next) ->
 
 LevelSessionSchema.statics.privateProperties = ['code', 'submittedCode', 'unsubscribed']
-LevelSessionSchema.statics.editableProperties = ['multiplayer', 'players', 'code', 'codeLanguage', 'completed', 'state',
+LevelSessionSchema.statics.editableProperties = ['multiplayer', 'players', 'code', 'codeLogs', 'codeLanguage', 'completed', 'state',
```
Should remove `codeLogs` from this list, that way we know that property only gets created on POSTing a code log.
